### ISSUE TITLE
Automatic Sync

### ISSUE THREAD

**Issue by mtbhuskies** [penge/my-notes#148]
Feature Suggestion: Have it automatically sync to Google Drive every X minutes (be extra nice to have that be a variable).

**Comment by penge** [comment]
There has been automatic sync on every open (and refresh) and close before. This has been turned off since the version 3.4.1, as there has been quite a number of requests back then, the other reason was more control. The good news is, that this could possibly be brought back as as an optional setting (also with additional interval like every X minutes, so technically it is possible). The other side to this is the quota for Google Drive API that comes for free, and beyond that is paid (need to look at the exact numbers and can provide here). As of now, it is without expenses but it could change if unlocked to be using much more requests multiplied by the number of users.

What I see as the next steps here is:

1. Check the exact quota for Google Drive API
2. See if we can do something about covering the running costs if quotas in point 1 are reached

**Comment by penge** [comment]
Who want's this? How important would this be for you? How about if it's optional in Options? Something like: "Enable Auto-sync".

**Comment by mtbhuskies** [comment]
+1 for me. An enable auto-sync would work great. I find myself forgetting to manually sync it, as I'm between multiple computers during the day (home pc, home laptop, work laptop)

**Comment by penge** [comment]
We'll have it.

**Comment by kevgo** [comment]
I find auto-sync when opening _my-notes_ extremely helpful. I use it on 5 different computers (work/private desktops, laptops) and often lose content because I forget to sync before starting to write, end when I sync later, new content disappears. 

I suggest an opt-in solution: add an option for this in the settings and leave it disabled by default. In my experience, only 1% of all users change such defaults. Hence, this optional feature should cause little extra traffic.

**Comment by penge** [comment]
@kevgo

Thanks for suggestion. Optional feature sounds best. Some might prefer it enabled, and some might actually prefer to sync manually (have it more under control).

Sync on open is a must. Sync on close could be there as well, something that would work even if you close the browser, but woudn't rely on that in situations where you put computer to sleep quickly or shut down.

Sync after every edit could be made too (like after 5 seconds of inactivity).

Sync button would still be there, so even if you don't type, you could always use it to sync with Google Drive.

How does that sound?

**Comment by kevgo** [comment]
Love it! I like the idea to sync after 5 seconds of inactivity much better than sync every 10 seconds no matter what. This way, there will be only "useful" traffic, i.e. updates with new content. And the "diffs" are smaller, which might prevent some sync conflicts. 

Sync at window close would be too unreliable for me to trust it  🙂 

Thanks for making this cool tool btw! I tried many browser plugins for note-taking, this is by far the best one. This is literally the last missing feature to make it perfect.

**Comment by penge** [comment]
@kevgo

Sorry for the longer message. Here's some news.

I have first prototype of this feature implemented. It does work. And it's available as optional feature in Options.

There are some things I decided to approach differently, and would like to hear your opinion about it.

First, or should I say last, I have decided to abandon auto sync at **window close**. It turned out to be unreliable as you said. Not only it is not guaranteed and for that sole reason shouldn't be there, I also run into CPU load problems. After opening several My Notes tabs and closing them at the same time (I know, this probably isn't a daily scenario but have to make that work anyway), the CPU run at 100%. The implementation approaches to do things on window close are vulnerable to memory leaks or CPU load problems. After the tabs were closed, CPU load was constantly at 100%. There was no place in service worker going in infinite loops (maybe some previous worker instance that was already detached). Disabling the extension helped in this case. This is a hard problem to identify. Then I decided, the best remedy for this, is just not do it. I looked at https://texteditor.co/ as my inspiration, and it also does not try to save things after window close. I hope it is fine not to have this unreliable part of the feature. I would say, it is better to not have.

Some good news next. Sync on **window open** is not an issue, works like a charm.

About Sync after 5 seconds of **inactivity** – I have decided to go a bit different here.
If you type a lot,  5 seconds of inactivity can be far from happening. That can lead to a potentially bigger text that is not having a backup in Google Drive.

Instead, I have decided to create a **timer** which checks every 6 seconds, if there is something new that needs to be synced. If there is, Auto Sync runs, if there is not, nothing happens. Using this approach, the most you can theoretically lose, are the changes in the past 6 seconds. The whole Auto Sync happens in the service worker, so the UI remains lean and fast. It happens only when the feature Auto Sync is enabled. And also, every open tab (if having My Notes open multiple times) will Auto Sync at once. When it comes to time, it can be 6 seconds, 10 seconds, any time we agree on. I found 6 seconds to be nice as the API used here works in minutes, and 6 seconds is there 0.1 minutes.

What do you think about these changes?

**Comment by kevgo** [comment]
This is great news! The 6-second interval for saving changes works for my use case and sounds like an overall solid solution. Thanks for the hard work on this, I can't wait to try this out!
